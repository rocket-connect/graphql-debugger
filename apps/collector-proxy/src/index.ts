import { Queue, QueueType } from "@graphql-debugger/queue";

import express, { Express } from "express";

import { postSchema } from "./routes/post-schema";
import { postTraces } from "./routes/post-traces";
import { postSchemaWorker } from "./workers/post-schema";
import { postTracesWorker } from "./workers/post-traces";

export const collector: Express = express();
collector.use(express.json({ limit: "500mb" }));

collector.post("/v1/traces", postTraces);
export const postTracesQueue = new Queue({
  type: QueueType.InMemory,
  worker: postTracesWorker,
});

collector.post("/v1/schema", postSchema);
export const postSchemaQueue = new Queue({
  type: QueueType.InMemory,
  worker: postSchemaWorker,
});
