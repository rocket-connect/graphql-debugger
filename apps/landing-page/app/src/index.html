<!doctype html>
<html lang="en">
  <head>
    <title>GraphQL Debugger</title>

    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
    <link
      rel="stylesheet"
      href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500;600;700&display=swap"
    />
  </head>

  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <!-- React app root element -->
    <div id="root"></div>
  </body>
</html>
