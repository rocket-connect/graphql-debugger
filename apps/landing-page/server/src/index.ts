import cors from "cors";
import express from "express";
import expressStaticGzip from "express-static-gzip";
import path from "path";

const config = {
  STATIC_FOLDER: path.join(__dirname, "../../app/build"),
  HTTP_PORT: 5000,
};

const app = express();
app.use(cors());
app.use(express.json());
app.use(expressStaticGzip(config.STATIC_FOLDER, {}));
app.use(express.static("public"));

app.get("*", expressStaticGzip(config.STATIC_FOLDER, {}));
app.use("*", expressStaticGzip(config.STATIC_FOLDER, {}));

async function main() {
  await app.listen(config.HTTP_PORT);

  console.log(`Listening at http://localhost:${config.HTTP_PORT}`);
}

main();
