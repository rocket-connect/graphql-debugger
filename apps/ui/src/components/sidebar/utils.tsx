import { History, HistoryActive as HistoryFilled } from "../../icons/history";
import { Star, StarFilled } from "../../icons/star";
import {
  configFilled,
  configStroke,
  folderFilled,
  folderStroke,
  loginFilled,
  loginStroke,
} from "../../images";
import { IDS } from "../../testing";
import { SchemaViewer } from "../schema/viewer";
import { Config } from "./views/config";
import { Favourites } from "./views/favourites";
import { History as HistoryView } from "./views/history";
import { Info } from "./views/info/info";
import { Login } from "./views/login";

interface SideBar {
  description: string;
  component: JSX.Element;
}

export const iconsMapper = (): Record<
  string,
  {
    active: React.ReactNode;
    inactive: React.ReactNode;
    hidden?: boolean;
    type: string;
    id: string;
  }
> => ({
  schema: {
    id: IDS.sidebar.icons.schemas,
    type: "schema",
    active: <img className="w-8" src={folderFilled} />,
    inactive: <img className="w-8" src={folderStroke} />,
  },
  config: {
    id: IDS.sidebar.icons.config,
    type: "config",
    active: <img className="w-8" src={configFilled} />,
    inactive: <img className="w-8" src={configStroke} />,
  },
  login: {
    id: IDS.sidebar.icons.login,
    type: "login",
    active: <img className="w-8" src={loginFilled} />,
    inactive: <img className="w-8" src={loginStroke} />,
  },
  history: {
    id: IDS.sidebar.icons.history,
    type: "history",
    active: <HistoryFilled size={"35"} color="#3b4b68" />,
    inactive: <History size={"35"} color="#3b4b68" />,
  },
  favourites: {
    id: IDS.sidebar.icons.favourites,
    type: "favourites",
    active: <Star size={"35"} color="#3b4b68" />,
    inactive: <StarFilled size={"35"} color="#3b4b68" />,
  },
});

export const sideBarComponentMapper = (
  currentSchema?: boolean,
): Record<string, SideBar> => ({
  schema: {
    description: currentSchema
      ? "Your GraphQL Schema, with analytics and traces"
      : "Select a GraphQL Schema to start exploring your traces",
    component: <SchemaViewer />,
  },
  info: {
    description: "",
    component: <Info />,
  },
  config: {
    description: "Configure GraphQL debugger settings, stored in your browser",
    component: <Config />,
  },
  login: {
    description:
      "Access GraphQL Debugger Cloud to view your projects and collaborate with your team",
    component: <Login />,
  },
  history: {
    description: "View your most recent traces",
    component: <HistoryView />,
  },
  favourites: {
    description: "View your favourite traces",
    component: <Favourites />,
  },
});

export const configs = [
  {
    name: "Dark Mode",
    description: "Toggle between light and dark mode",
    callout: "(Comming Soon)",
    enabled: false,
  },
  {
    name: "Cookies",
    description: "We collect data to improve your experience",
    enabled: true,
    alwaysEnabled: true,
  },
];
