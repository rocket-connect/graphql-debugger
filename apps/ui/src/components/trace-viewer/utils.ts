import { Span } from "@graphql-debugger/types";

export type RenderTree = Omit<Span, "__typename"> & {
  children: RenderTree[];
};

export const createTreeData = (spanArray: Span[]): RenderTree[] => {
  const treeData: RenderTree[] = [];
  const lookup: { [key: string]: RenderTree } = {};

  spanArray.forEach((span) => {
    lookup[span.spanId] = {
      spanId: span.spanId,
      name: span.name,
      traceId: span.traceId,
      children: [],
      startTimeUnixNano: span.startTimeUnixNano,
      endTimeUnixNano: span.endTimeUnixNano,
      durationNano: span.durationNano,
      parentSpanId: span.parentSpanId as string,
      createdAt: span.createdAt,
      kind: span.kind,
      id: span.id,
      updatedAt: span.updatedAt,
      errorMessage: span.errorMessage,
      errorStack: span.errorStack,
    };
  });

  spanArray.forEach((span) => {
    if (span.parentSpanId) {
      lookup[span.parentSpanId]?.children?.push(lookup[span.spanId]);
    } else {
      treeData.push(lookup[span.spanId]);
    }
  });

  return treeData;
};
