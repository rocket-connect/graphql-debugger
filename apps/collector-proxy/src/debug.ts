import { Debug, type Debugger } from "@graphql-debugger/utils";

const debugNamespace = "collector-proxy";

export const debug: Debugger = Debug(debugNamespace);
