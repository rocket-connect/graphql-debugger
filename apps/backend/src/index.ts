import { DebuggerClient } from "@graphql-debugger/client";
import { createServer } from "@graphql-debugger/graphql-schema";
import { SpanExporter, SpanProcessor } from "@graphql-debugger/opentelemetry";
import { graphqlDebugger } from "@graphql-debugger/plugin-express";

import cors from "cors";
import express, { Express } from "express";
import expressStaticGzip from "express-static-gzip";
import http from "http";
import path from "path";

import { TRACE_EXPRESS } from "./config";
import { debug } from "./debug";

export async function start({
  port,
  client,
  spanProcessorFactory,
}: {
  port: string;
  client: DebuggerClient;
  spanProcessorFactory?: (exporter: SpanExporter) => SpanProcessor;
}) {
  try {
    const app: Express = express();
    app.use(cors());
    app.use(express.json());
    if (TRACE_EXPRESS) {
      app.use(graphqlDebugger());
    }
    app.use(
      "/graphql",
      createServer({
        client,
        spanProcessorFactory,
      }),
    );
    app.use(
      "/",
      expressStaticGzip(path.join(__dirname, "../../ui/build"), {
        enableBrotli: true,
      }),
    );
    app.use(express.static("public"));

    const server = await app.listen(port);

    debug("Application started");

    return server;
  } catch (error) {
    debug("Application failed to start", error);
    throw error;
  }
}

export async function stop({ server }: { server: http.Server }) {
  await server.close();
}

export * from "./config";
