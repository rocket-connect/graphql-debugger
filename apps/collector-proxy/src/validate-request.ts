import { z } from "@graphql-debugger/schemas";

import { NextFunction, Request, Response } from "express";

import { debug } from "./debug";

export function validateRequest(schema: z.AnyZodObject) {
  return async (req: Request, res: Response, next: NextFunction) => {
    debug(`Request ${req.method} ${req.url}`);

    try {
      for (const key in req.body) {
        if (["body", "headers"].includes(key)) {
          try {
            JSON.parse(req.body[key] || "{}");
          } catch (error) {
            return res
              .status(400)
              .send(`body.${key} is not a valid JSON string`)
              .end();
          }
        }
      }

      await schema.parseAsync({
        body: req.body,
        query: req.query,
        params: req.params,
      });

      return next();
    } catch (error) {
      const e = error as Error;

      return res.status(400).json({
        message: e.message,
      });
    }
  };
}
