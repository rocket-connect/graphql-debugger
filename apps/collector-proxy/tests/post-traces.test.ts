import { TRACER_NAME } from "@graphql-debugger/opentelemetry";
import {
  AttributeNames,
  PostTraces,
  ResourceSpans,
} from "@graphql-debugger/types";

import { faker } from "@faker-js/faker";
import { describe, expect, test } from "@jest/globals";
import { parse, print } from "graphql";
import util from "util";

import { client } from "../src/client";
import { request } from "./utils";

const sleep = util.promisify(setTimeout);

describe("POST /v1/traces", () => {
  test("should throw when no body is sent", async () => {
    const response = await request().post("/v1/traces").send({});

    expect(response.status).toBe(400);

    const body = await response.body;

    expect(body.message).toMatchSnapshot();
  });

  test("should throw span validation error when not sent correctly", async () => {
    const knownScope: ResourceSpans["scopeSpans"][0]["scope"] = {
      name: TRACER_NAME,
    };

    const payload: PostTraces["body"] = {
      resourceSpans: [
        {
          scopeSpans: [
            {
              scope: knownScope,
              spans: [
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore - testing the validation error on a span
                {
                  spanId: "1",
                },
              ],
            },
          ],
        },
      ],
    };

    const response = await request().post("/v1/traces").send(payload);

    expect(response.status).toBe(400);

    const body = await response.body;

    expect(body.message).toMatchSnapshot();
  });

  test("should receive traces and map them correctly in the database", async () => {
    const knownScope: ResourceSpans["scopeSpans"][0]["scope"] = {
      name: TRACER_NAME,
    };

    const schemaHash = faker.string.alpha(6);

    const document = /* GraphQL */ `
      query {
        users {
          name
          age
          posts {
            title
            content
            comments {
              content
            }
          }
        }
      }
    `;

    const payload: PostTraces["body"] = {
      resourceSpans: [
        {
          scopeSpans: [
            {
              scope: knownScope,
              spans: [
                {
                  spanId: "2a5f8b696abf9858",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: undefined,
                  name: "query users",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.OPERATION_NAME,
                      value: {
                        stringValue: "users",
                      },
                    },
                    {
                      key: AttributeNames.OPERATION_TYPE,
                      value: {
                        stringValue: "query",
                      },
                    },
                    {
                      key: AttributeNames.DOCUMENT,
                      value: {
                        stringValue: document,
                      },
                    },
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                    {
                      key: AttributeNames.OPERATION_RETURN_TYPE,
                      value: {
                        stringValue: "[User]",
                      },
                    },
                    {
                      key: AttributeNames.OPERATION_ROOT,
                      value: {
                        boolValue: true,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 1, low: 1 },
                  endTimeUnixNano: { high: 1, low: 1 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
                {
                  spanId: "b496e68882240e49",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "2a5f8b696abf9858",
                  name: "User name",
                  kind: 0,
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 2, low: 2 },
                  endTimeUnixNano: { high: 2, low: 2 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                },
                {
                  spanId: "443fd2e5e4f00845",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "2a5f8b696abf9858",
                  name: "User age",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 3, low: 3 },
                  endTimeUnixNano: { high: 3, low: 3 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
                {
                  spanId: "c96662d3273a7415",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "2a5f8b696abf9858",
                  name: "User posts",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 4, low: 4 },
                  endTimeUnixNano: { high: 4, low: 4 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
                {
                  spanId: "4aa8f7495bc7546c",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "c96662d3273a7415",
                  name: "Posts title",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 5, low: 5 },
                  endTimeUnixNano: { high: 5, low: 5 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
                {
                  spanId: "b8b21753292a8794",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "c96662d3273a7415",
                  name: "Posts content",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 6, low: 6 },
                  endTimeUnixNano: { high: 6, low: 6 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
                {
                  spanId: "3877c153fa0d43d6",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "c96662d3273a7415",
                  name: "Posts comments",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 7, low: 7 },
                  endTimeUnixNano: { high: 7, low: 7 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
                {
                  spanId: "9c027ebcb7fb03a3",
                  traceId: "46cb720c4cc8b0c1e28cabd112057b78",
                  parentSpanId: "3877c153fa0d43d6",
                  name: "Comment content",
                  kind: 0,
                  attributes: [
                    {
                      key: AttributeNames.SCHEMA_HASH,
                      value: {
                        stringValue: schemaHash,
                      },
                    },
                  ],
                  droppedAttributesCount: 0,
                  startTimeUnixNano: { high: 8, low: 8 },
                  endTimeUnixNano: { high: 8, low: 8 },
                  droppedEventsCount: 0,
                  droppedLinksCount: 0,
                  events: [],
                  links: [],
                  status: {
                    code: 0,
                  },
                },
              ],
            },
          ],
        },
      ],
    };

    const response = await request().post("/v1/traces").send(payload);

    expect(response.status).toBe(200);

    await sleep(2000); // backoff the writes using sqlite

    const traceGroup = await client.trace.findFirst({
      where: {
        traceId: payload.resourceSpans[0].scopeSpans[0].spans[0].traceId,
      },
      options: {
        includeSpans: true,
      },
    });

    expect(traceGroup).toBeDefined();

    expect(traceGroup?.spans.length).toEqual(8);
    traceGroup?.spans.forEach((span) => {
      expect(span.isForeign).toEqual(false);
      expect(span.graphqlSchemaHash).toEqual(schemaHash);

      if (span.isGraphQLRootSpan) {
        expect(span.attributes).toBeTruthy();
        expect(span.graphqlDocument).toBeTruthy();

        const stringAttributes = span.attributes as string;
        const parsedAttributes = JSON.parse(stringAttributes);

        expect(parsedAttributes).toMatchObject({
          [AttributeNames.OPERATION_NAME]: "users",
          [AttributeNames.OPERATION_TYPE]: "query",
          [AttributeNames.OPERATION_RETURN_TYPE]: "[User]",
          [AttributeNames.OPERATION_ROOT]: true,
        });
      }
    });

    const rootSpan = traceGroup?.spans.find((span) => span.isGraphQLRootSpan);
    expect(rootSpan).toBeDefined();
    expect(rootSpan?.name).toEqual("query users");
    expect(rootSpan?.graphqlDocument).toEqual(print(parse(document)));
  });
});
