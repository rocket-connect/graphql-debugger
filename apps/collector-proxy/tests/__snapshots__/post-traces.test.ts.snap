// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /v1/traces should throw span validation error when not sent correctly 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "traceId"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "name"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "kind"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "startTimeUnixNano"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "endTimeUnixNano"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "attributes"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans",
      0,
      "scopeSpans",
      0,
      "spans",
      0,
      "status"
    ],
    "message": "Required"
  }
]"
`;

exports[`POST /v1/traces should throw when no body is sent 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "body",
      "resourceSpans"
    ],
    "message": "Required"
  }
]"
`;
